addon-api/common/Self.js

```diff
@@ -13,7 +13,56 @@ export default class Self extends Listenable {
     this.disabled = false;
     this.addEventListener("disabled", () => (this.disabled = true));
     this.addEventListener("reenabled", () => (this.disabled = false));
+    this.addEventListener("message", (event) => {
+      const data = {
+        UserscriptAddon: "user",
+        BackgroundScriptAddon: "persistent",
+      };
+      if ((event.detail.scope || []).includes(data[this._addonObj.constructor.name])) {
+        for (const func of this._msgFunctions) {
+          func.call(this, event.detail.msg);
+        }
+      }
+    });
+    this._msgFunctions = [];
   }
+  sendMessage(msg, data = {}) {
+    const { addonId, scope } = data;
+    const sending = {
+      target: "self",
+      name: "message",
+      addonId,
+      data: { msg, scope },
+    };
+    if (this._template) {
+      // CS
+      this._template.setAttribute(`data-fire-event__${Date.now()}`, JSON.stringify(sending));
+    } else {
+      // PS
+      chrome.tabs.query({}, (tabs) =>
```

Comment:
But... this wouldn't be available in firefox... would it?

Reply:
Firefox supports most `chrome.*` APIs.

Reply:
We use this same function in other places as well.